torch/_inductor/fx_passes/group_fusion.py

```diff
@@ -0,0 +1,188 @@
+import logging
+import operator
+from typing import List
+
+import networkx
+import torch
+from torch._dynamo.utils import counters
+
+from ..pattern_matcher import (
+    CallFunctionVarArgs,
+    config_flag,
+    get_arg_value,
+    get_mutation_region_id,
+    stable_topological_sort,
+)
+
+log = logging.getLogger(__name__)
+
+
+def get_nx_graph_from_fx_graph(graph: torch.fx.Graph) -> networkx.DiGraph:
+    G = networkx.DiGraph()
+    for node in graph.nodes:
+        G.add_node(node)
+        for user in node.users:
+            G.add_edge(node, user)
+    return G
+
+
+class GroupFusionPass:
+    """
+    `GroupFusionPass` implements an Fx pass which groups together similar nodes into a fused op. Unlike `PatternMatcherPass`
+    which can only match local patterns in a graph, this pass can group together nodes across the graph. The only restriction
+    imposed is that the nodes being fused must not have a path in between each other (which would lead to a cycle). Additional
+    checks such as `extra_check` (at a node level), or `pair_check` can be added to pose additional restrictions on
+    nodes to be fused.
+    """
+
+    def __init__(
+        self,
+        pattern,
+        pair_check,
+        replacement_fn,
+        extra_check=lambda m: True,
+        prevent_match_across_mutations=False,
```

Comment:
I worry people will forget to set this.  I think there is some could to automatically detect if we should use this.  Maybe @williamwen42 can provide a pointer.

Or we could make it required arg.

Reply:
Made it a required arg for now

Reply:
This isn't a required arg (you can remove the default for that)